reactions/src/SPT.cpp

```diff
@@ -633,58 +632,68 @@ real SPT::Lag_bias(int a, real k, real bias[]) const {
 /////// Numerical 1-loop spectra /////////////
 
 static double ploopn2_mgdd( const PowerSpectrum& P_L, double vars[], double k, double r, double x){
-  double kargs[4],kv[3],xv[3], p22,p13,d;
+  double kargs[4],kv[3],xv[3], p22,p13;
+  // tolerance for ode solver (see SpecialFunctions.cpp, initn2). This encounters a singularity if k'.-k' = -1 exactly.
+  double tol = 1e-8;
+  // The integrated |k'| is parametrised as k*r
   IOW iow;
         kv[0] = k;
         kv[1] = k*r;
         kv[2] = kv[1];
-        xv[0] = -0.99999999; // don't set to -1 otherwise solver encounters singularity
+        xv[0] = -1. + tol;
         xv[1] = x;
         xv[2] = -x;
         kargs[0] = sqrt(kv[1]*kv[1]+kv[0]*kv[0]-2.*kv[1]*kv[0]*xv[1]);
         kargs[2] = sqrt(kv[1]*kv[1]+2.*kv[1]*kv[2]*xv[0]+kv[2]*kv[2]);
         kargs[1] = sqrt(kv[2]*kv[2]+2.*kv[2]*kv[0]*xv[2]+kv[0]*kv[0]);
         kargs[3] = sqrt(kv[0]*kv[0]+2.*kv[0]*kv[1]*xv[1]+kv[1]*kv[1]);
-        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4]);
+        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4],vars[6]);
         p22 = pow2(F2_nk);
         p13 = F1_nk*F3_nk;
 
     return pow2(r)*2.*P_L(k*r)*(P_L(kargs[0])*p22 + 3.*P_L(k)*p13);
 }
 
+
 static double ploopn2_mgdt( const PowerSpectrum& P_L, double vars[], double k, double r, double x){
-  double kargs[4],kv[3],xv[3], p22,p13,d;
+  double kargs[4],kv[3],xv[3], p22,p13;
+  // tolerance for ode solver (see SpecialFunctions.cpp, initn2). This encounters a singularity if k'.-k' = -1 exactly.
+  double tol = 1e-8;
+  // The integrated |k'| is parametrised as k*r
   IOW iow;
         kv[0] = k;
         kv[1] = k*r;
         kv[2] = kv[1];
-        xv[0] = -0.99999999;
+        xv[0] = -1. + tol ;
         xv[1] = x;
         xv[2] = -x;
         kargs[0] = sqrt(kv[1]*kv[1]+kv[0]*kv[0]-2.*kv[1]*kv[0]*xv[1]);
         kargs[2] = sqrt(kv[1]*kv[1]+2.*kv[1]*kv[2]*xv[0]+kv[2]*kv[2]);
         kargs[1] = sqrt(kv[2]*kv[2]+2.*kv[2]*kv[0]*xv[2]+kv[0]*kv[0]);
         kargs[3] = sqrt(kv[0]*kv[0]+2.*kv[0]*kv[1]*xv[1]+kv[1]*kv[1]);
-        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4]);
+        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4],0.);
```

Comment:
Why do we use Omega_nu=0 here instead of vars[6] as in ploopn2_mgdd?

Reply:
We don't use the velocity spectra currently so left total matter as a source, but agree for consistency we should change that.